server/logging/firehose.go

```diff
@@ -0,0 +1,141 @@
+package logging
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/firehose"
+	"github.com/aws/aws-sdk-go/service/firehose/firehoseiface"
+	"github.com/pkg/errors"
+)
+
+const (
+	firehoseMaxRetries = 3
+
+	// See
+	// https://docs.aws.amazon.com/sdk-for-go/api/service/firehose/#Firehose.PutRecordBatch
+	// for documentation on limits.
+	firehoseMaxRecordsInBatch = 500
+	firehoseMaxSizeOfRecord   = 1000 * 1000     // 1,000 KB
+	firehoseMaxSizeOfBatch    = 4 * 1000 * 1000 // 4 MB
+)
+
+type firehoseLogWriter struct {
+	client firehoseiface.FirehoseAPI
+	stream string
+}
+
+func NewFirehoseLogWriter(region, id, secret, stream string) (*firehoseLogWriter, error) {
+	sess, err := session.NewSession(&aws.Config{
+		Credentials: credentials.NewStaticCredentials(id, secret, ""),
+		Region:      &region,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "create Firehose client")
+	}
+	client := firehose.New(sess)
+
+	f := &firehoseLogWriter{client, stream}
+	if err := f.validateStream(); err != nil {
+		return nil, errors.Wrap(err, "create Firehose writer")
+	}
+	return f, nil
+}
+
+func (f *firehoseLogWriter) validateStream() error {
+	out, err := f.client.DescribeDeliveryStream(
+		&firehose.DescribeDeliveryStreamInput{
+			DeliveryStreamName: &f.stream,
+		},
+	)
+	if err != nil {
+		return errors.Wrapf(err, "describe stream %s", f.stream)
+	}
+
+	if (*(*out.DeliveryStreamDescription).DeliveryStreamStatus) != firehose.DeliveryStreamStatusActive {
+		return errors.Errorf("delivery stream %s not active", f.stream)
+	}
+
+	return nil
+}
+
+func (f *firehoseLogWriter) Write(logs []json.RawMessage) error {
+	var records []*firehose.Record
+	totalBytes := 0
+	for _, log := range logs {
+		if len(log) > firehoseMaxSizeOfRecord {
+			// TODO zwass: better logging
```

Comment:
Yeah... this is a hard limit for Firehose and it mimics the behavior of osquery in the case of logs that are too big. At the least I'd like to make this log to stderr some information about the log that is dropped. Maybe it could write oversize logs to a local file?

Reply:
>  At the least I'd like to make this log to stderr some information about the log that is dropped. Maybe it could write oversize logs to a local file?

A stderr message sounds good enough. I think logging them to a local file brings up concerns around rotating that file (which you will have to consider because you are in fact writing massive logs to it).

At the end of the day, the goal is for end-users to easily identify the queries generating massive logs. If I were a user and I had result logs that were oversized, I'd want to see the info about the scheduled query producing the large logs and the actual size of the logs in both bytes and in # of rows. I'd also want to see the host UUID any decorators so I could identify the host easily.

Reply:
I'm with Jason. If the logs are so big they can't go to firehose, storing them on disk is going to raise all kinds of rotation questions. Logging enough for debugging ought be enough.